### PR TITLE
Fix meson baremetal

### PR DESCRIPTION
**Pull request by jhol**
Changelog: Omit
Docs: Omit

~Bugfix: #16902~
Bugfix: #16903

- [X] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

**Comment by memsharded** [comment]
Thanks for your contribution

This is breaking a test:

<details>
<pre>
________________________ test_baremetal_cross_compiling ________________________

[gw2] linux -- Python 3.12.2 /home/conan/.pyenv/versions/3.12.2/bin/python



    @pytest.mark.tool("meson")

    def test_baremetal_cross_compiling():

        meson_build = textwrap.dedent("""

            project('tutorial', 'cpp')

            executable('demo', 'main.cpp')

            """)

        main_cpp = gen_function_cpp(name="main")

        profile_baremetal = textwrap.dedent("""

            [settings]

            arch=x86_64

            os=baremetal

            compiler=gcc

            compiler.cppstd=gnu17

            compiler.libcxx=libstdc++11

            compiler.version=13

            build_type=Release

            """)

    

        client = TestClient()

        client.save({"conanfile.py": _conanfile_py,

                     "meson.build": meson_build,

                     "main.cpp": main_cpp,

                     "baremetal": profile_baremetal})

        client.run("install . --profile:build=default --profile:host=baremetal")

>       client.run("build .")



test/functional/toolchains/meson/test_cross_compilation.py:273: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

conan/test/utils/tools.py:599: in run

    return self._run_cli(command_line, assert_error=assert_error)

conan/test/utils/tools.py:573: in _run_cli

    self._handle_cli_result(command_line, assert_error=assert_error, error=error, trace=trace)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 



self = <conan.test.utils.tools.TestClient object at 0x7fe50ace1160>

command = 'build .', assert_error = False, error = 1

trace = 'Traceback (most recent call last):\n  File "/home/jenkins/workspace/ConanTestSuitev2_PR-16904/conans/errors.py", line...tion: conanfile.py: Error in build() method, line 28\n\tmeson.configure()\n\tConanException: Error 1 while executing\n'



    def _handle_cli_result(self, command, assert_error, error, trace=None):

        if (assert_error and not error) or (not assert_error and error):

            if assert_error:

                msg = " Command succeeded (failure expected): "

            else:

                msg = " Command failed (unexpectedly): "

            exc_message = "\n{header}\n{cmd}\n{output_header}\n{output}\n".format(

                header='{:=^80}'.format(msg),

                output_header='{:=^80}'.format(" Output: "),

                cmd=command,

                output=str(self.stderr) + str(self.stdout) + "\n" + str(self.out)

            )

            if trace:

                exc_message += '{:=^80}'.format(" Traceback: ") + f"\n{trace}"

>           raise Exception(exc_message)

E           Exception: 

E           ======================== Command failed (unexpectedly): ========================

E           build .

E           =================================== Output: ====================================

E           

E           ======== Input profiles ========

E           Profile host:

E           [settings]

E           arch=x86_64

E           build_type=Release

E           compiler=gcc

E           compiler.libcxx=libstdc++11

E           compiler.version=8

E           os=Linux

E           

E           Profile build:

E           [settings]

E           arch=x86_64

E           build_type=Release

E           compiler=gcc

E           compiler.libcxx=libstdc++11

E           compiler.version=8

E           os=Linux

E           

E           

E           ======== Computing dependency graph ========

E           Graph root

E               conanfile.py: /tmp/tmpb0jvxwbxconans/path with spaces/conanfile.py

E           

E           ======== Computing necessary packages ========

E           

E           ======== Installing packages ========

E           

E           ======== Installing packages ========

E           

E           ======== Finalizing install (deploy, generators) ========

E           conanfile.py: Calling generate()

E           conanfile.py: Generators folder: /tmp/tmpb0jvxwbxconans/path with spaces

E           conanfile.py: Generating aggregated env files

E           conanfile.py: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']

E           

E           ======== Calling build() ========

E           conanfile.py: Calling build()

E           conanfile.py: Meson configure cmd: meson setup --cross-file "/tmp/tmpb0jvxwbxconans/path with spaces/conan_meson_cross.ini" --native-file "/tmp/tmpb0jvxwbxconans/path with spaces/conan_meson_native.ini" "/tmp/tmpb0jvxwbxconans/path with spaces/build" "/tmp/tmpb0jvxwbxconans/path with spaces" --prefix=/

E           conanfile.py: RUN: meson setup --cross-file "/tmp/tmpb0jvxwbxconans/path with spaces/conan_meson_cross.ini" --native-file "/tmp/tmpb0jvxwbxconans/path with spaces/conan_meson_native.ini" "/tmp/tmpb0jvxwbxconans/path with spaces/build" "/tmp/tmpb0jvxwbxconans/path with spaces" --prefix=/

E           The Meson build system

E           Version: 1.5.1

E           Source dir: /tmp/tmpb0jvxwbxconans/path with spaces

E           Build dir: /tmp/tmpb0jvxwbxconans/path with spaces/build

E           Build type: cross build

E           Project name: tutorial

E           Project version: undefined

E           

E           ../meson.build:2:0: ERROR: 'cpp' compiler binary not defined in cross file [binaries] section

E           

E           A full log can be found at /tmp/tmpb0jvxwbxconans/path with spaces/build/meson-logs/meson-log.txt

E           

E           ERROR: conanfile.py: Error in build() method, line 28

E           	meson.configure()

E           	ConanException: Error 1 while executing
</pre>
</details>

File: test/functional/toolchains/meson/test_cross_compilation.py [hunk 1]
```diff
@@ -244,3 +244,34 @@ def test_use_meson_toolchain():
     # Build locally
     c.run('build . --profile:host=android')
     assert "conanfile.py (hello/0.1): Calling build()" in c.out
+
+
+@pytest.mark.tool("meson")
+def test_baremetal_cross_compiling():
```
Comment (memsharded) [inline test/functional/toolchains/meson/test_cross_compilation.py]: Most likely this test should be restricted to Linux only, the configuration for ``gcc`` in other plaforms require more configuration.

**Review by memsharded — APPROVED** [review]
Thanks for your fix @jhol 

I have moved the test to "integration" and simplified it, so it is faster (as it is not building anything, it is important to keep the test suite fast, and it seems enough to cover the change), but still doing the same check of ``system = "none"`` in the toolchain file (by the way, how did you know it? I have checked the meson docs, but it is not evident what are the values for ``system``, do you have any link?

Let's try to get this in the next Conan 2.8 release.

**Review by franramirez688 — APPROVED** [review]
LGTM

**Comment by memsharded** [comment]
Merged, if you still have any feedback about the test change @jhol, please let me know.

Thanks again for your contribution!

**Comment by memsharded** [comment]
Uhm...

Doing an extra search, I have found this in Meson codebase: https://github.com/mesonbuild/meson/blob/18427adbf21909f66a307a54ba4c47fd91e18fba/cross/ccrx.txt#L19

It seems that ``system = "bare metal"`` is a valid value? Maybe also the previous one "baremetal"?
https://github.com/search?q=repo%3Amesonbuild%2Fmeson+bare+metal&type=code

**Comment by franramirez688** [comment]
> Uhm...
> 
> Doing an extra search, I have found this in Meson codebase: https://github.com/mesonbuild/meson/blob/18427adbf21909f66a307a54ba4c47fd91e18fba/cross/ccrx.txt#L19
> 
> It seems that `system = "bare metal"` is a valid value? Maybe also the previous one "baremetal"? https://github.com/search?q=repo%3Amesonbuild%2Fmeson+bare+metal&type=code

As far as I can see, the official list is this one https://mesonbuild.com/Reference-tables.html#operating-system-names but you can read on the bottom of the box says:

```
Any string not listed above is not guaranteed to remain stable in future releases.
```
Bare metal does not appear there so it seems there is no official way to list it.
However, the kernel names (https://mesonbuild.com/Reference-tables.html#kernel-names-since-120) list it and it's named as `none`.

**Comment by jhol** [comment]
I just tested compiling picolibc with this patch non-applied, and it seems to build fine now - due to other fixes, so I think the patch could be reverted. I'm sorry for the confusion caused.

**Comment by memsharded** [comment]
Reverting this in https://github.com/conan-io/conan/pull/16929

**Pull request merged**